### PR TITLE
fix(import): validate duplicate identifiers in student_update import (#641)

### LINKED ISSUE #641: Student Update CSV import bypasses duplicate identifier validation
## Problem

The `student_update` CSV import can overwrite a student with a Student ID, APAAR ID, or PEN number that already belongs to another student. The normal create/update flow validates identifier conflicts, but the import path calls the raw update service directly and skips that validation.

This happened on production in import `1574` (1,265 rows), which recreated 10 duplicate APAAR groups after a cleanup.

## Current path

- `ImportWorker.process_student_update_record/1` finds the student.
- It calls `StudentUpdateService.update_student_with_user_data/2`.
- That calls `Users.update_student_with_user/3` without the shared duplicate-identifier validation.

The new database unique indexes will stop duplicate values from being stored, but the import currently halts on the first constraint error after earlier rows may already have been updated.

## Expected

- Validate Student ID, APAAR ID, and PEN conflicts before updating each row.
- Report the conflicting row and identifier clearly.
- Define and test import behavior so a duplicate does not produce an unclear partial import.
- Add regression coverage for duplicate identifiers in `student_update` imports.

Related rollout PR: #581

### PR DESCRIPTION
**Pull request by Bahugunajii**
Fixes #641.

## Problem
The `student_update` CSV import called the raw update service directly, bypassing the duplicate-identifier validation used by the normal create/update flow. A row could move another student's **APAAR ID** (or **PEN number**) onto a different student — recreating duplicate APAAR groups on prod (import 1574, 1,265 rows). Only the DB unique index caught it, surfacing as a raw `Ecto.Changeset` interpolated into a string and halting the run with an unclear message.

## Fix
- Add **`Users.validate_identifier_conflicts/2`** — a pre-write check that mirrors the student table's global unique constraints and returns a clear, human-readable message (`"APAAR ID '…' already exists for another student"`). Only identifiers present in the row are checked; the student being updated is excluded (re-saving its own identifiers is allowed).
- **`StudentUpdateService.update_student_with_user_data/2`** runs the check before writing, so it covers both the CSV import and `update_student_by_student_id/2`.
- The import framework already prepends the CSV row number, so a conflict is persisted as `%{row: N, error: "APAAR ID '…' already exists for another student"}` and the offending row fails cleanly instead of overwriting another student.

### Import behavior
Kept the existing fail-fast model (a row error halts the run), now with a clear, row-attributed message. Rows before the conflict stay applied; the operator fixes the sheet and re-runs. No change to the global import loop.

### Why `student_id` is not checked here
The merged auth-group-scoping work makes `student_id` unique only **within an auth group** (see `create_or_update_student/1`). A global `student_id` check would falsely reject an update to any student whose id is legitimately reused in another auth group (the update row carries the student's own id). So the guard covers only the genuinely-global identifiers the DB enforces: **`apaar_id`, `pen_number`**. (PEN is not yet an importable CSV column; the guard covers it at the service layer for when it is, and for direct service calls.)

## Tests
- Service-level: rejects moving another student's `apaar_id`/`pen_number`; allows re-saving own identifiers; allows a new non-conflicting identifier.
- Users context: unit tests for `validate_identifier_conflicts/2`, including that `student_id` is **not** flagged.
- End-to-end: first `student_update` worker test — a conflicting APAAR row halts the import with a clear error and leaves the other student untouched (reproduces the prod incident).

All new tests pass; `mix format`, `mix credo`, and `mix compile --warnings-as-errors` are clean.

## ⚠️ Pre-existing `main` issues (NOT introduced by this PR, not fixed here)
1. **Duplicate migration version `20260701120000`** — shared by `add_lms_student_ingestion.exs` and `add_user_session_indexes.exs`. This makes `mix ecto.migrate` fail on a fresh DB, so **CI on this PR will be red until it is resolved on `main`** (e.g. renumber one migration). Independent of this change.
2. **Global `student_id` unique index vs auth-group scoping** — the LMS-ingestion migration creates `student_student_id_unique_not_null` (never dropped in any `up`), while the auth-group-scoping feature expects `student_id` not to be globally unique. As a result, 3 tests in `users_test.exs` ("create_or_update_student/1 auth-group scoping") fail on clean `main`. Independent of this change.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: lib/dbservice/users.ex [hunk 1]
```diff
@@ -691,6 +691,48 @@ defmodule Dbservice.Users do
...
+
+  Only identifiers that are present and non-empty in `params` are checked, so a partial
+  update touching unrelated fields passes untouched. Accepts string or atom keys. When
+  `existing_student` is given, that student is excluded from the check (re-saving its
+  own identifiers is allowed); pass `nil` when validating a brand new student.
+
+  Returns `:ok`, or `{:error, message}` naming the conflicting identifier.
+  """
+  def validate_identifier_conflicts(params, existing_student \\ nil) do
+    apaar_id = identifier_param(params, :apaar_id)
```
Comment (deepansh96) [inline lib/dbservice/users.ex:780]: Blocking: please trim the APAAR ID before checking and saving it. I tested this locally with an existing APAAR ID `222222222222`, then updated another Student using ` 222222222222 ` with spaces. The update succeeded, and both values were stored. This means spaces can bypass the duplicate check. Please trim the value, treat a whitespace-only value as empty, and add a CSV test for this case.

File: lib/dbservice/users.ex [hunk 2]
```diff
@@ -691,6 +691,48 @@ defmodule Dbservice.Users do
...
+    pen_number = identifier_param(params, :pen_number)
```
Comment (deepansh96) [inline lib/dbservice/users.ex:781]: Question: the new check handles `pen_number`, but I could not find a PEN column or mapping in the `student_update` CSV. Can this CSV path currently receive a PEN value? Should this PR do one of these two things?

1. Add PEN to the CSV mapping and add an import test for a duplicate PEN.
2. Keep PEN as service-level protection only and clearly say that PEN is outside the CSV change in this PR.

I want to confirm which behavior we expect before merging.

Comment (Bahugunajii) [reply lib/dbservice/users.ex:781]: Will add PEN to the mapping file

File: lib/dbservice/users.ex (hunk 1 above)
Comment (Bahugunajii) [reply lib/dbservice/users.ex:780]: Fixed. apaar_id and pen_number are now trimmed on both sides: the `Student` changeset trims them (whitespace-only → nil), and `validate_identifier_conflicts/2` normalizes the incoming value the same way before comparing — so a padded ` 222222222222 ` maps to the stored `222222222222` and the conflict is caught. Added a CSV worker test (padded APAAR halts the import) plus unit tests for the trim and the whitespace-only case.

File: lib/dbservice/users.ex (hunk 2 above)
Comment (Bahugunajii) [reply lib/dbservice/users.ex:781]: Went with option 1. Added `student_pen_number` to the `student`/`student_update` CSV mapping, and added an import test that halts when a row moves another student's PEN onto a different student. Updated the PR description accordingly.

**Comment by Bahugunajii** [comment]
@suryabulusu on both points: (1) trimmed the apaar (and pen) from the sheet — padded values can no longer bypass the duplicate check. (2) You're right the pre-existing issues are fixed on main — I merged latest main into the branch, which resolves the duplicate migration version (`add_user_session_indexes` renamed to `20260724110000`) and drops the global `student_id` unique index/constraint. Updated the description to reflect this. A fresh full-suite run now matches main's baseline exactly (21 pre-existing failures, 0 in the identifier area).

**Review by deepansh96 — APPROVED** [review]
Rechecked the latest fixes locally. The focused APAAR/PEN update and import tests pass.
